docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -82,87 +91,125 @@ This code is extremely ugly, but leads to functional virtual function calls on t
 ![VPointerToDevice](./figures/VirtualFunctions-VPointerToDevice.png)
 
 Note that like with other uses of `new`, you need to later `free` the memory.
+
+## Make it portable
+
+We can get rid of the call to `cudaMalloc()` and make the previous code portable
+
+```cpp
+auto deviceInstanceMemory = Kokkos::kokkos_malloc(sizeof(Derived)); // allocate memory on device
+Kokkos::parallel_for("Initialize", 1, KOKKOS_LAMBDA (const int i) {
+  new (static_cast<Derived*>(deviceInstanceMemory)) Derived(); // initialize on device
+});
+Kokkos::fence("Wait for initialize");
+Base* deviceInstance = static_cast<Derived*>(deviceInstanceMemory); // declare on this memory
+
+Kokkos::parallel_for("DeviceKernel", SomeCudaPolicy, KOKKOS_LAMBDA (const int i) {
+  // use on device
+});
+Kokkos::fence();
+
+Kokkos::parallel_for("Destroy", 1, KOKKOS_LAMBDA (const int i) {
+  deviceInstance->~Base(); // destroy on device
+});
+Kokkos::fence("Wait for destroy");
+Kokkos::kokkos_free(deviceInstanceMemory); // free
+```
+
+We replaced `cudaMalloc()` by `Kokkos::kokkos_malloc()` and introduced a distinction between the memory that the instance uses, and the actual instance object.
+Since the kernel does not have a return type, we use a static cast to associate the later to the former.
+Then, the object is destroyed, and its memory freed.
+
```

Comment:
I guess if we put the kokkos version we should do it all around and get rid of the cuda version

Reply:
I'm not sure to get what you mean.

Reply:
I meant update all the code sections to use kokkos semantics and get rid of the cuda code.

Reply:
Ah, yes indeed.

Reply:
Done in the latest commits.